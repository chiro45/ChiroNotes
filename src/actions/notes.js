import Swal from "sweetalert2";


import { db } from "../firebase/firebaseConfing";

import { types } from "../types/types";

import {loadNotes} from '../helpers/loadNotes'
import { fileUpload } from "../helpers/fileUpload";




export const startNewNote = ()=>{
    return async (dispatch, getState)=>{       //el getState nos trae todo el state de la aplicacion
        const {uid} = getState().auth;   
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        const doc =  await db.collection(`${uid}/journal/notes`).add(newNote)     
        
        dispatch(activeNote(doc.id, newNote))

        dispatch(addNewNote(doc.id, newNote))
    }
}

export const activeNote = (id,note )=>({
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
});

export const addNewNote = (id, note)=>({
        type: types.notesAddNew,
        payload:{
            id,
            ...note
        }
})


export const startLoadingNotes = (uid) =>{
    return async (dispatch)=>{
        const notes = await loadNotes(uid)
      
        dispatch(setNotes(notes))
    }
}
export const setNotes = (notes)=>({
    type: types.notesLoad,
    payload: notes
})


 export const startSaveNote = (note)=>{
    return async(dispatch, getState)=>{
        try {
            const {uid} = getState().auth;
        if(!note.url){
            delete note.url;
        }
        const noteToFirestore ={...note}

        delete noteToFirestore.id;

        if(note.title.length > 1){
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        dispatch(refreshNote(note.id,note ))

        Swal.fire('Guardado!', note.title, "success")
        }else{
            Swal.fire('Deberia tener un titulo!', note.title, "error")
        }

        } catch (error) {
            Swal.fire('Error en el guardado!',"hubo un problema, intenta nuevamente", "error")
        }


    }
}

export const refreshNote = (id, note) =>({
    type: types.notesUpdated,
    payload:{
        id,
        note:{
                id,
                ...note
        }

        
    }
})


export const startUploading =(file) =>{
    return async(dispatch, getState)=>{
        const {active:activeNote} = getState().notes

        Swal.fire({
            title: 'Subiendo...',
            text: 'Por Favor Espere...',
            allowOutsideClick: false,
            didOpen:()=>{
                Swal.showLoading()
            }
            
        });
       
        const fileUrl = await fileUpload(file)
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))

        Swal.close();
    }
}


export const startDeleting = (id)=>{
    return async(dispatch, getState)=>{
        
        const uid = getState().auth.uid;

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id))

        Swal.fire('Eliminado!','', "success")
    }
}


export const deleteNote = (id)=>({
    type: types.notesDelete,
    payload: id
})


export const noteLogout = () =>({
    type: types.notesLogoutCleaning
   
})
