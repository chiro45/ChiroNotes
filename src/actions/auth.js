import Swal from 'sweetalert2'


import { types } from "../types/types"
import {firebase, googleAuthProvider} from '../firebase/firebaseConfing'
import { finishLoading, startLoading } from "./ui"
import { noteLogout} from './notes'
//acciones que son usadas en el auth




export const startLoginEmailPassword = (email,password)=>{
    return(dispatch)=>{
        
        dispatch(startLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
        .then( ({ user }) =>{


            dispatch(login(user.uid, user.displayName))
            
            
        dispatch(finishLoading())

        })
        .catch(e => {
        Swal.fire("Error","El usuario o contraseÃ±a son incorrectos","error")
    })
}}
export const startRegisterWhithEmailPasswordName = (email,password, name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
        .then( async({ user })=>{

            await user.updateProfile({
                displayName: name

            })

            dispatch(
                login(user.uid, user.displayName)
             )
            
         
        })
        .catch(e => {
            Swal.fire("Error","El usuario ya ha sido registrado","error")
        
    })

}}


export const startGoogleLogin = ()=>{
    return (dispatch)=>{
       firebase.auth().signInWithPopup(googleAuthProvider) 
       .then(({ user })=>{

           dispatch(
               login(user.uid, user.displayName, user.photoURL),

               console.log(user.photoURL)
                 
               )
            })
            
       
       
       
    }
}


export const login = (uid, displayName, photoURL)=>{
        return{
            type: types.login,
            payload: {
                uid,
                photoURL,
                displayName,
            }
        }
}


export const startLogout = () =>{
    return async(dispatch)=>{
        firebase.auth().signOut()

        dispatch(logout())
        
        dispatch(noteLogout())
    }
}

export const logout = () =>({
    type: types.logout
})


