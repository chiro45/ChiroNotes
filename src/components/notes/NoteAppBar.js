import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';
import moment from 'moment'
export const NoteAppBar = ({date}) => {
    const {active:note} = useSelector(state => state.notes)
    const dispatch = useDispatch();

    const handleSave =()=>{
        dispatch(startSaveNote(note))
    }

    const handlePictureClick= () =>{
        document.getElementById('fileSelector').click()




    }


    const handleFileChange = (e)=>{
            const file = e.target.files[0];
            if (file){
                dispatch(startUploading(file))
            }
    }

    
    
  return (
    <div className='notes__appbar'>
  
        <span>{moment().subtract(10, 'days').calendar()}</span>
        <input id='fileSelector'
               type='file'
               name='file'
               style={{display: 'none'}}
               onChange={handleFileChange}
        />
        <div>
            <button className='btn'
            onClick={handlePictureClick}>Foto</button>
        
            <button className='btn'
            onClick={handleSave}>Guardar</button>
        </div>
    

    </div>)
};
