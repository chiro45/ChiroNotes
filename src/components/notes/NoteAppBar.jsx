
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';
import moment from 'moment'

export const NoteAppBar = ({date}) => {
    const {active:note} = useSelector(state => state.notes)
    
    
    //declaramos el dispatch
    const dispatch = useDispatch();
    //evento que dispara el gurdarNote
    const handleSave =()=>{
        dispatch(startSaveNote(note))
    }
    //obtenemos el bottton
    const handlePictureClick= () =>{
        document.getElementById('fileSelector').click()
    }

    
    //evento que detecta que cambie
    const handleFileChange = (e)=>{
            const file = e.target.files[0];
            //si el file existe emtonces se produce el disparo de actualizacion
            if (file){
                dispatch(startUploading(file))
            }
    }

    
    
  return (
    <div className='notes__appbar'>
        
        <div><span className='appBar'> <i className="fas fa-calendar-day"></i>{moment().format('L')}</span></div>
        <div>
        <input id='fileSelector'
               type='file'
               name='file'
               style={{display: 'none'}}
               onChange={handleFileChange}
        />
            <button className='btn appBar'
            onClick={handlePictureClick}>Foto</button>
        
            <button className='btn appBar'
            onClick={handleSave}>Guardar</button>
        </div>
    

    </div>)
};
