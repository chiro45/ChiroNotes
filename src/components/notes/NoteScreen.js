import React, { useEffect, useRef } from 'react';
import {  useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';

import { NoteAppBar } from './NoteAppBar';

export const NoteScreen = () => {


  const {active:note} = useSelector(state => state.notes)

  const dispatch = useDispatch();
  
  const activeId = useRef(note.id)

  const [formValues, handleInputChange, reset]=useForm(note)

  useEffect(()=>{
     
   if(note.id !== activeId.current){
     reset(note)
     activeId.current = note.id
   }

  },[note, reset])
 
  useEffect(()=>{
     
    dispatch(activeNote(formValues.id, {...formValues}))
 
   },[formValues , dispatch])


  const {body, title, id} = formValues;


  const handleDelete=()=>{

    
    dispatch(startDeleting(id))
    
  }



  return (
    <div className='notes__main-content animate__animated animate__fadeInRight animate__fast'>
        <NoteAppBar/>

        <div className='notes__content '>
                <input
                type='text'
                placeholder='Agrega un titulo'
                className='notes__title-input'
                autoComplete='off'
                value={title}
                name='title'
                onChange={handleInputChange}
                ></input>
                <textarea
                placeholder='QuÃ© paso hoy?'
                className='notes__textarea'
                value={body}
                name='body'
                onChange={handleInputChange}
                  >
                </textarea>

               {
                  (note.url) &&
                  <div className='notes__image'>
                  <img
                  src={note.url}
                  alt='imagen'

                  ></img>
                </div>
                } 

        </div>
     
          <button 
          className='btn btn-danger'
          onClick={handleDelete}
          >Borrar Nota</button>
        
    </div>
  );
};
