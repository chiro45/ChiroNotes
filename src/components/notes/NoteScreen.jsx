import  { useEffect, useRef } from 'react';
import {  useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';

import { NoteAppBar } from './NoteAppBar';

export const NoteScreen = () => {

  //extraemos el active del state 
  const {active:note} = useSelector(state => state.notes)
  // declaramos el dispatch
  const dispatch = useDispatch();
  //nos retorna un el id mutable
  const activeId = useRef(note.id)
  //utilizamos el useForm y mandamos la nota como el intialState
  const [formValues, handleInputChange, reset]=useForm(note)
  //el useEffect nos 
  useEffect(()=>{
          //si el id de la nota es diferente al actual
        if(note.id !== activeId.current){
          //resetea la nota
          reset(note)
          //y le cambiamos el id
          activeId.current = note.id
        }

  },[note, reset])
  //cuando se produzca un cambio en el formValues
  useEffect(()=>{
     //se produce el 
    dispatch(activeNote(formValues.id, {...formValues}))
 
   },[formValues , dispatch])

   //desestructuramos los datos del formValue
  const {body, title, id} = formValues;


  const handleDelete=()=>{

    
    dispatch(startDeleting(id))
    
  }



  return (
    <div className='notes__main-content animate__animated animate__fadeInRight animate__fast'>
        <NoteAppBar/>

        <div className='notes__content '>
                <input
                type='text'
                placeholder='Agrega un titulo'
                className='notes__title-input'
                autoComplete='off'
                value={title}
                name='title'
                onChange={handleInputChange}
                ></input>
                <textarea
                placeholder='QuÃ© paso hoy?'
                className='notes__textarea'
                value={body}
                name='body'
                onChange={handleInputChange}
                  >
                </textarea>

               {
                  (note.url) &&
                  <div className='notes__image'>
                  <img
                  src={note.url}
                  alt='imagen'

                  ></img>
                </div>
                } 

        </div>
     

        <div className='journal__containerButton'>
          <button 
          className='btn btn-danger'
          onClick={handleDelete}
          >Borrar Nota</button>
        </div>
    </div>
  );
};
