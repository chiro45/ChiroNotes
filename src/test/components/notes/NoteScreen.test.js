import { mount } from "enzyme";
import { Provider } from "react-redux";
import { NoteScreen } from "../../../components/notes/NoteScreen";
import '@testing-library/jest-dom'
import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);


jest.mock('../../../actions/notes',()=>({
    activeNote: jest.fn()
}))

const initState = {
    auth: {
        uid:'123123',
        name:'luciano',
        photoURL: null
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{
        active: {
            id: 1234,
            title: 'title',
            body: 'body',
            date: 0
        },
        notes:[]
        
    }
};
const store = mockStore(initState)
store.dispatch= jest.fn();


describe('Pruebas en el <NoteScreen/>', () => {
    
    const wrapper = mount(
        <Provider store={store}>
        <NoteScreen/>
        </Provider>
    )
    
    test('debe de mostrarse correctamente', () => {
        
        expect(wrapper).toMatchSnapshot()

    });


    test('debe de disparar el active note', () => {
      
        wrapper.find('input[name="title"]').simulate('change',{
            target: {
                name: 'title',
                value: 'hola de nuevo'
            }
        })

        expect(activeNote).toHaveBeenLastCalledWith(
            1234, {
                "body": "body",
                 "date": 0, "id": 1234,
                  "title": "hola de nuevo"
            }
            )

    });
    
    
});
