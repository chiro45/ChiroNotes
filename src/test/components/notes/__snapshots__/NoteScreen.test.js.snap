// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <NoteScreen/> debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeInRight animate__fast"
>
  <div
    className="notes__appbar"
  >
    <div
      className="notes__nav-toggle"
      onClick={[Function]}
    >
      <span
        className="nav-toggle appBar"
        id="navT"
      >
        <i
          className=" fas fa-times "
        />
      </span>
      <span
        className="nav-toggle appBar hidden"
        id="navT2"
      >
        <i
          className="fas fa-bars"
        />
      </span>
    </div>
    <span
      className="appBar"
    >
       
      <i
        className="fas fa-calendar-day"
      />
        
      02/05/2022
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn appBar"
        onClick={[Function]}
      >
        Foto
      </button>
      <button
        className="btn appBar"
        onClick={[Function]}
      >
        Guardar
      </button>
    </div>
  </div>
  <div
    className="notes__content "
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Agrega un titulo"
      type="text"
      value="title"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="QuÃ© paso hoy?"
      value="body"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar Nota
  </button>
</div>
`;
